spatialos-sdk/src/worker/worker_future.rs

```diff
@@ -0,0 +1,95 @@
+use futures::channel::oneshot::*;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    thread,
+    thread::JoinHandle,
+};
+
+pub struct WorkerFuture<T: WorkerSdkFuture> {
+    inner: WorkerFutureState<T>,
+}
+
+impl<T: WorkerSdkFuture> WorkerFuture<T> {
+    pub fn new(future: T) -> Self {
+        WorkerFuture {
+            inner: WorkerFutureState::NotStarted(future),
+        }
+    }
+}
+
+enum WorkerFutureState<T: WorkerSdkFuture> {
+    NotStarted(T),
+    InProgress(InProgressHandle<T>),
+}
+
+struct InProgressHandle<T: WorkerSdkFuture> {
+    pub result_rx: Receiver<T::Output>,
+    pub thread: Option<JoinHandle<()>>,
+}
+
+impl<T: WorkerSdkFuture> Drop for WorkerFuture<T> {
+    fn drop(&mut self) {
+        if let WorkerFutureState::InProgress(handle) = &mut self.inner {
+            handle.thread.take().unwrap().join().unwrap();
+        }
+    }
+}
+
+struct SendPtr<T> {
+    pub ptr: *mut T,
+}
+
+unsafe impl<T> Send for SendPtr<T> {}
```

Comment:
We can tidy this up a little bit, but we may still be sending pointers between the threads. For example, to `start` a worker connection future which connects via the Locator, we need to pass the locator across the thread boundaries (which is just a thin wrapper around the `Worker_Locator*`). 

https://github.com/jamiebrynes7/spatialos-sdk-rs/pull/134/files/4e78dfa3692a968d268cee27693c094f54dd3cc7#diff-9b16ecc341594f655ad8daad5e018ff2R635-R638

This is probably still preferable to the current implementation (which is a bit of a hack). I think the `Locator` doesn't have `Clone` implemented and requires you to move it into the future, so that seems a lot safer

Reply:
Yeah, under the hood we're absolutely just passing raw pointers between threads, but we could be leaning more heavily on Rust's type system to enforce the invariants necessary to make that safe. Especially when it comes to dealing with threads, leaning on Rust's ownership semantics is probably the best way to make sure we're not introducing any data races.

Reply:
Refactored - PTAL :smile: